Support for Numpy integers and fixed slicing

swapsha96 opened the pull request:
In issue #299, I've discussed that Lightcurve does not support Numpy integers like `int64`. This is a simple fix for that.

Another issue is with slicing. If `new_lc = lc[:]` is run, the length of Lightcurve counts reduces by half. Since I am not an astronomer, I am not familiar with 'Good Time Intervals' and 'Time Resolution'. So I would need help to fix the second problem.

**Needs discussion**: If we could create a `copy()` function so that the package provides more options other than slicing.

pep8speaks commented:
Hello @swapsha96! Thanks for submitting the PR.

 - In the file [`stingray/lightcurve.py`](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py), following are the PEP8 issues :

> [Line 16:80](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py#L16): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 79 characters)
> [Line 211:5](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py#L211): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 272:80](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py#L272): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 312:80](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py#L312): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 429:20](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py#L429): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 433:80](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py#L433): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 79 characters)
> [Line 446:80](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py#L446): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 447:80](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py#L447): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (97 > 79 characters)
> [Line 599:9](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py#L599): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 608:18](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py#L608): [E272](https://duckduckgo.com/?q=pep8%20E272) multiple spaces before keyword
> [Line 676:12](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py#L676): [E201](https://duckduckgo.com/?q=pep8%20E201) whitespace after '('
> [Line 676:39](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py#L676): [E202](https://duckduckgo.com/?q=pep8%20E202) whitespace before ')'
> [Line 697:13](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py#L697): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 706:13](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py#L706): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 716:38](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py#L716): [E711](https://duckduckgo.com/?q=pep8%20E711) comparison to None should be 'if cond is not None:'
> [Line 716:47](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py#L716): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 716:48](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py#L716): [E262](https://duckduckgo.com/?q=pep8%20E262) inline comment should start with '# '
> [Line 718:78](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py#L718): [E262](https://duckduckgo.com/?q=pep8%20E262) inline comment should start with '# '
> [Line 718:80](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py#L718): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 719:49](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py#L719): [E201](https://duckduckgo.com/?q=pep8%20E201) whitespace after '('
> [Line 719:80](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py#L719): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (110 > 79 characters)
> [Line 719:102](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py#L719): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 738:80](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py#L738): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 751:5](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py#L751): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 869:80](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py#L869): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (123 > 79 characters)
> [Line 869:104](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py#L869): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ','
> [Line 941:80](https://github.com/StingraySoftware/stingray/blob/e86cd68bd11fa41d5fa6748954b8cc4e18e66a1f/stingray/lightcurve.py#L941): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)

matteobachetti commented on stingray/lightcurve.py:
```diff
@@ -1059,7 +1059,7 @@ def plot(self, witherrors=False, labels=None, axis=None, title=None,
             plt.title(title)
 
-        if save:
+        if save is True:
```
Just `if save` is sufficient. If one writes `save='mickeymouse'` when the docs say that `save` is `bool`... well... whatever happens, they had it coming :)

swapsha96 replied:
Done. :+1:

matteobachetti commented:
@swapsha96 can you add a test in `test_lightcurve.py` where you make the same slices that were failing in the notebook and show that this new mechanism works? Thanks!

swapsha96 commented:
There's obviously a mistake in slicing. If `step` is greater than 1 then it doesn't work. As I mentioned earlier, since I am not an astronomer, I am not familiar with 'Good Time Intervals' and 'Time Resolution'. So I would need help to fix the second problem.

matteobachetti commented:
@swapsha96 I don't see the failures you mention. Tests should include these cases and fail until you make the necessary modifications.

matteobachetti commented:
About GTIs: they are the intervals in the time series when the instrument is working correctly and the observing conditions are good. We need to only run the analysis during these intervals, or the periodicities we measure are incorrect.
The time resolution is just the binning time of the time series: in a light curve, we count the photons that are caught every, e.g., 0.5 seconds. This is the time resolution.

matteobachetti commented on stingray/lightcurve.py:
```diff
@@ -418,8 +418,7 @@ def __getitem__(self, index):
                 new_gti = cross_two_gtis(new_gti, new_gt1)
-
-            new_gti = cross_two_gtis(self.gti, new_gti)
+                new_gti = cross_two_gtis(self.gti, new_gti)
```
Why did you indent here?

swapsha96 replied:
I observed that putting that line inside `if` block *helps* to pass the test. I have been too busy in last few days that I couldn't work on it. I'll fix it. Even the test is not correctly copied from the notebook. Sorry about that. I'll work on it.

swapsha96 commented:
@matteobachetti I've removed the intent and modified the test function (copied the snippet directly from the Lightcurve tutorials notebook). It fails because the size reduces by half.

matteobachetti commented:
@swapsha96 let's wait until #307 is merged, then you can rebase and see if the latest modifications are sufficient to make the test pass

matteobachetti commented:
@swapsha96 ok, #307 is merged. Let's rebase! Thanks

swapsha96 commented:
Following test that I wrote still failed:-

![screenshot from 2018-02-07 09-14-41](https://user-images.githubusercontent.com/14021461/35897710-9118380e-0be8-11e8-8e50-13c110aedc30.jpg)

matteobachetti commented:
@swapsha96 I made a PR to your fork, fixing the problem (I think)

swapsha96 commented:
@matteobachetti All Lightcurve tests passed and the notebook looks good too.
![screenshot from 2018-02-08 18-22-48](https://user-images.githubusercontent.com/14021461/35974072-b01d04e4-0cfd-11e8-81ae-e23e70aee542.png)
Are we good to go? :smile:

matteobachetti commented:
Thanks @swapsha96!

swapsha96 commented:
@matteobachetti Couldn't have finished it without you. Thanks! :sweat_smile:
